torch/lib/c10d/TcpStore.cpp

```diff
@@ -0,0 +1,274 @@
+    std::string key = tcputil::recvString(socket);
+    int64_t addVal = tcputil::recvValue<int64_t>(socket);
+
+    if (tcpStore_.find(key) != tcpStore_.end()) {
+      auto buf = reinterpret_cast<const char*>(tcpStore_[key].data());
+      auto len = tcpStore_[key].size();
+      addVal += std::stoll(std::string(buf, len));
+    }
+    auto addValStr = std::to_string(addVal);
+    tcpStore_[key] = std::vector<uint8_t>(addValStr.begin(), addValStr.end());
+    // Now send the new value
+    tcputil::sendValue<int64_t>(socket, addVal);
+    // On "add", wake up all of the processes that wait
+    // for keys already in the store
+    wakeUpWaitingRanks(key);
+
+  } else if (qt == QueryType::GET) {
+    std::string key = tcputil::recvString(socket);
+    auto data = tcpStore_.at(key);
+    tcputil::sendVector<uint8_t>(socket, data);
+
+  } else if (qt == QueryType::CHECK) {
+    SizeType nargs;
+    tcputil::recvBytes<SizeType>(socket, &nargs, 1);
+    std::vector<std::string> keys(nargs);
+    for (size_t i = 0; i < nargs; i++) {
+      keys[i] = tcputil::recvString(socket);
+    }
+    // Now we have received all the keys
+    if (checkAndUpdate(keys)) {
+      tcputil::sendValue<QueryType>(socket, QueryType::STOP_WAITING);
+    } else {
+      for (auto& key : keys) {
+        waiting_[key].push_back(rank);
+      }
+      keysAwaited_[rank] = keys.size();
+      tcputil::sendValue<QueryType>(socket, QueryType::KEEP_WAITING);
+    }
+  } else {
+    throw std::runtime_error("expected a query type");
+  }
+}
+
+bool TcpStoreDaemon::checkAndUpdate(std::vector<std::string>& keys) const {
+  bool ret = true;
+  for (auto it = keys.begin(); it != keys.end();) {
+    if (tcpStore_.count(*it) == 0) {
+      ret = false;
+      it++;
+    } else {
+      it = keys.erase(it);
+    }
+  }
+  return ret;
+}
+
+// TcpStore class methods
+
+TcpStore::TcpStore(const std::string& masterAddr,
+                   PortType masterPort,
+                   bool isServer)
+ : isServer_(isServer)
+ , tcpStoreAddr_(masterAddr)
+ , tcpStorePort_(masterPort)
+
+{
+  if (isServer_) {
+    // Openning up the listening socket
+    std::tie(masterListenSocket_, std::ignore) = tcputil::listen(masterPort);
+    // Now start the daemon
+    tcpStoreDaemon_ = std::unique_ptr<TcpStoreDaemon>(
+        new TcpStoreDaemon(masterListenSocket_)
+    );
+  }
+  // Connect to the daemon
+  storeSocket_ = tcputil::connect(tcpStoreAddr_, tcpStorePort_);
+}
+
+TcpStore::~TcpStore() {
+  ::close(storeSocket_);
+  if (isServer_) {
+    ::close(masterListenSocket_);
+    /**
+     * Store daemon should end because of closed connection.
+     * daemon destructor should join the thread
+     */
+    tcpStoreDaemon_.reset(nullptr);
+  }
+}
+
+void TcpStore::set(const std::string& key, const std::vector<uint8_t>& data) {
+  tcputil::sendValue<QueryType>(storeSocket_, QueryType::SET);
+  tcputil::sendString(storeSocket_, key, true);
+  tcputil::sendVector<uint8_t>(storeSocket_, data);
```

Comment:
What do you think of making every command request/response to make the interaction pattern predictable?